File: lib/src/main/java/net/mm2d/upnp/internal/server/Address.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.server;

import net.mm2d.util.NetworkUtils;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.UnknownHostException;

import javax.annotation.Nonnull;

/**
 * マルチキャストアドレス。
 */
public enum Address {
    /**
     * IPv4用マルチキャストアドレス
     */
    IP_V4("239.255.255.250"),
    /**
     * IPv6用リンクローカルマルチキャストアドレス
     */
    IP_V6_LINK_LOCAL("FF02::C"),
    /**
     * IPv6用サイトローカルマルチキャストアドレス
     *
     * <p>未使用
     */
    @SuppressWarnings("unused")
    IP_V6_SITE_LOCAL("FF05::C"),
    ;

    @Nonnull
    private final InetAddress mInetAddress;
    @Nonnull
    private final InetSocketAddress mInetSocketAddress;

    Address(@Nonnull final String address) {
        mInetAddress = parseAddress(address);
        mInetSocketAddress = new InetSocketAddress(mInetAddress, SsdpServer.SSDP_PORT);
    }

    @Nonnull
    InetAddress getInetAddress() {
        return mInetAddress;
    }

    @Nonnull
    InetSocketAddress getSocketAddress() {
        return mInetSocketAddress;
    }

    @Nonnull
    public String getAddressString() {
        return NetworkUtils.getAddressString(mInetSocketAddress);
    }

    static InetAddress parseAddress(@Nonnull final String address) {
        try {
            return InetAddress.getByName(address);
        } catch (final UnknownHostException e) {
            throw new AssertionError(e);
        }
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/server/Address.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.server

import net.mm2d.upnp.common.util.toAddressString
import java.net.InetAddress
import java.net.InetSocketAddress

/**
 * Multicast address
 */
internal enum class Address(
    ssdpAddress: String,
    eventAddress: String
) {
    /**
     * Multicast address for IPv4
     */
    IP_V4(
        ServerConst.SSDP_ADDRESS_V4,
        ServerConst.EVENT_ADDRESS_V4
    ),
    /**
     * Multicast address for IPv6 (link local)
     */
    IP_V6(
        ServerConst.SSDP_ADDRESS_V6,
        ServerConst.EVENT_ADDRESS_V6
    ),
    ;

    val ssdpInetAddress: InetAddress = InetAddress.getByName(ssdpAddress)
    val ssdpSocketAddress: InetSocketAddress = InetSocketAddress(ssdpInetAddress, ServerConst.SSDP_PORT)
    val ssdpAddressString: String = ssdpSocketAddress.toAddressString()
    val eventInetAddress: InetAddress = InetAddress.getByName(eventAddress)
}
